File: MAP/lab4/src/models/fileHandling/OpenRFile.java
package models.fileHandling;

import exceptions.fileExceptions.FileException;
import models.PrgState;
import models.statement.Statement;
import utils.IDictionary;

import java.io.*;


public class OpenRFile implements Statement {

    private String fileName;
    private String varName;

    public OpenRFile(String varName, String fileName) {
        this.fileName = fileName;
        this.varName = varName;
    }

    @Override
    public PrgState execute(PrgState state) {

        IFileTable<Integer, FileData> fileTable = state.getFileTable();
        IDictionary<String, Integer> dict = state.getDict();
        BufferedReader reader;

        for(FileData d: fileTable.getAllValues()) {
            if (d.getFileName().equals(fileName))
                throw new FileException("The file is already opened in FileTable");
        }

        try{
            reader = new BufferedReader(new FileReader(fileName));
         } catch (IOException e) {
             throw new FileException("File not found");
         }

         int id = GeneratorInteger.gen_ID();

         fileTable.add(id, new FileData(fileName, reader));
         dict.setValue(varName, id);

         return state;
    }

    @Override
    public String toString(){
           return "open("+varName+", "+fileName+')';
    }

}

File: MAP/lab7/lab7/Models/FileHandlers/OpenRFile.cs
using System;
using System.IO;
using System.Runtime.CompilerServices;

namespace lab7.Models.FileHandlers
{
    public class OpenRFile : IStatement
    {
        private string varName;
        private string fname;
        private static int id = 0;

        public OpenRFile(string varName, string fname)
        {
            this.varName = varName;
            this.fname = fname;
        }

        public PrgState execute(PrgState state)
        {
            foreach (FileData i in state.FileTable.allValues()) 
                if(i.Name == fname)
                    throw new SystemException("this file " + fname + " is already opened");

            if(!File.Exists(fname))
                throw new SystemException("this file " + fname + " can't be found");
            
            
            StreamReader sw = new StreamReader(fname, true);
            state.FileTable.add(id, new FileData(fname, sw));
            state.Dict.setValue(varName, id++); 

            return state;

        }

        public override string ToString()
        {
            return "Open(" + fname + ")";

        }
    }
}